### DISCUSSION TITLE
Custom Controller to handle incoming request

### DISCUSSION THREAD

**Discussion by SparshPareek** [fastapi/fastapi#8374] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import gzip
from typing import Callable, List
import time

from fastapi import Body, FastAPI, Request, Response
from fastapi.routing import APIRoute

app = FastAPI()

@app.get("/urlinfo/cat")
async def root(url: str):
    return {"message": "Categroy"}

@app.get("/urlinfo/rep")
async def root(url: str):
    return {"message": "Reputation"}

@app.get("/urlinfo/topcat")
async def root(url: str):
    return {"message": "Top Category"}
```

### Description

Hi @tiangolo,

For my college project I want to create a custom route handler which would take the incoming request and assign handlers based on the request url path.

Example:
I have below endpoints:
/urlinfo/cat --> provides the site category
/urlinfo/rep --> provides the site reputation
/urlinfo/topcat --> provides top categories

Now above endpoints have a common query parameter url: /urlinfo/cat?url="", /urlinfo/rep?url="".
As the endpoints have same query parameters can we implement something like /urlinfo/cat/rep?url="". By this it would provide both the category and the reputation for the site. Now there are multiple endpoints under urlinfo which has same query parameter so would like to implement a solution like custom request routing handler or controller which would first check the url and call there respective endpoints get the response from all the endpoints and finally merge all responses together and return the same.

### Operating System

Linux, Windows, macOS

### FastAPI Version

0.79.1

### Python Version

3.10.6

**Comment by JarroVGIT** [comment]
This could be handled by using a `path` classifier in a `Path()` parameter. The following gives you an idea how this could work (the functions that are called could just as well be endpoint functions!). Also, in your example all your endpoint path operations functions are called `root()`, which will cause issues.

This example will run as-is.

```python
from collections import defaultdict

from fastapi import FastAPI, Query

app = FastAPI()


def cat(url: str):
    return {"message": "category"}


def rep(url: str):
    return {"message": "Reputation"}


def topcat(url: str):
    return {"message": "Top Category"}


mapping = {"cat": cat, "rep": rep, "topcat": topcat}


@app.get("/urlinfo/{to_include:path}")
def get_urlinfo(to_include: str, url: str = Query(None)):
    kinds = set(to_include.strip("/").split("/"))
    responses = defaultdict(list)
    for kind in kinds:
        func = mapping.get(kind, None)
        if func:
            for key, value in (func(url)).items():
                responses[key].append(value)
    return responses


@app.get("/")
async def root():
    return {"hello": "world"}


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=8000)
```

Using curl:

```bash
% curl "localhost:8000/urlinfo/cat/topcat?url=blaat"
{"message":["category","Top Category"]}%             

% curl "localhost:8000/urlinfo/topcat/topcat/top?url=blaat"
{"message":["Top Category"]}%    

curl "localhost:8000/urlinfo/topcat/rep/?url=blaat"
{"message":["Top Category","Reputation"]}%

**Comment by SparshPareek** [comment]
Thanks @JarroVGIT, the solution works fine; although i was thinking if there’s a way to override APIRouter or Middleware utilising the incoming request somehow to achieve the functionality.

**Comment by JarroVGIT** [comment, marked as answer]
You probably could, but I have a hard time to understand what the benefit of that approach would be over the approach given using `{param:path}`, especially since it is a much better use case for it. 

Note that having an arbitrary amount of query parameters is highly advised over using an arbitrary amount of path parameters in API's, as they get to be incredibly unintuitive very quickly. So, in your case, it would make a lot more sense to create your endpoints like `localhost:8000/urlinfo/{url}?kind=rep&kind=topcat`. That way, you can build your logic around the query parameters, rather than deciphering an arbitrary amount of path parameters.
